HidPkg/Crates/HiiKeyboardLayout/src/lib.rs

```diff
@@ -0,0 +1,775 @@
+  type Error = scroll::Error;
+  fn try_from_ctx(src: &'_ [u8], _ctx: scroll::Endian) -> Result<(Self, usize), Self::Error> {
+    let offset = &mut 0;
+    let efi_key =
+      EfiKey::try_from(src.gread::<u32>(offset)?).map_err(|err| scroll::Error::BadInput { size: 0, msg: err })?;
+    Ok((efi_key, *offset))
+  }
+}
+
+impl ctx::TryIntoCtx<scroll::Endian> for &EfiKey {
+  type Error = scroll::Error;
+  fn try_into_ctx(self, dest: &mut [u8], _ctx: scroll::Endian) -> Result<usize, Self::Error> {
+    let offset = &mut 0;
+    dest.gwrite(*self as u32, offset)?;
+    Ok(*offset)
+  }
+}
+
+// Convenience macro for defining HiiKey::Key structures.
+macro_rules! key {
+  ($key:expr, $unicode:literal, $shifted:literal, $alt_gr:literal, $shifted_alt_gr:literal, $modifier:expr, $affected:expr ) => {
+    HiiKey::Key(key_descriptor!($key, $unicode, $shifted, $alt_gr, $shifted_alt_gr, $modifier, $affected))
+  };
+}
+
+// convenience macro for defining HiiKeyDescriptor structures.
+// note: for unicode characters, these are encoded as u16 for compliance with UEFI spec. UEFI only supports UCS-2
+// encoding - so unicode characters that require more than two bytes under UTF-16 are not supported (and will panic).
+macro_rules! key_descriptor {
+  ($key:expr, $unicode:literal, $shifted:literal, $alt_gr:literal, $shifted_alt_gr:literal, $modifier:expr, $affected:expr ) => {
+    HiiKeyDescriptor {
+      key: $key,
+      unicode: $unicode.encode_utf16(&mut [0u16; 1])[0],
+      shifted_unicode: $shifted.encode_utf16(&mut [0u16; 1])[0],
+      alt_gr_unicode: $alt_gr.encode_utf16(&mut [0u16; 1])[0],
+      shifted_alt_gr_unicode: $shifted_alt_gr.encode_utf16(&mut [0u16; 1])[0],
+      modifier: $modifier,
+      affected_attribute: $affected,
+    }
+  };
+}
+
+/// Returns a default HiiKeyboardLayout (which is a standard US-104 layout)
+#[rustfmt::skip]
+pub fn get_default_keyboard_layout() -> HiiKeyboardLayout {
+  HiiKeyboardLayout {
+    guid: efi::Guid::from_fields(0x3a4d7a7c, 0x18a, 0x4b42, 0x81, 0xb3, &[0xdc, 0x10, 0xe3, 0xb5, 0x91, 0xbd]),
+    keys: vec![
+      key!(EfiKey::C1,          'a',    'A',    '\0',   '\0',   NULL_MODIFIER,                AFFECTED_BY_STANDARD_SHIFT | AFFECTED_BY_CAPS_LOCK),
+      key!(EfiKey::B5,          'b',    'B',    '\0',   '\0',   NULL_MODIFIER,                AFFECTED_BY_STANDARD_SHIFT | AFFECTED_BY_CAPS_LOCK),
+      key!(EfiKey::B3,          'c',    'C',    '\0',   '\0',   NULL_MODIFIER,                AFFECTED_BY_STANDARD_SHIFT | AFFECTED_BY_CAPS_LOCK),
+      key!(EfiKey::C3,          'd',    'D',    '\0',   '\0',   NULL_MODIFIER,                AFFECTED_BY_STANDARD_SHIFT | AFFECTED_BY_CAPS_LOCK),
+      key!(EfiKey::D3,          'e',    'E',    '\0',   '\0',   NULL_MODIFIER,                AFFECTED_BY_STANDARD_SHIFT | AFFECTED_BY_CAPS_LOCK),
+      key!(EfiKey::C4,          'f',    'F',    '\0',   '\0',   NULL_MODIFIER,                AFFECTED_BY_STANDARD_SHIFT | AFFECTED_BY_CAPS_LOCK),

```

Comment:
I turned off the formatter because by default it will put every argument to key!() on its own line, which is anarchy. But I will fix the formatting issues you identified.

Reply:
Fixed.